corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,924 @@
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    # parameter for plot
+    parm1 = -0.5*np.log10(kgain)
+    
+    # Gaussian read noise value in DN
+    mean_rn_gauss_DN = np.mean(rn_gauss)
+    mean_rn_gauss_e = mean_rn_gauss_DN * kgain
+    
+    mean_rn_std_DN = np.mean(read_noise)
+    mean_rn_std_e = mean_rn_std_DN * kgain
+    
+    # If requested, plotting
+    if make_plot:
+        fname = 'kgain_ptc'
+        # Create log-spaced averages for plotting
+        full_range_averages = np.logspace(logspace_start, logspace_stop, logspace_num)
+        
+        plt.figure(num=3, figsize=(20, 10))
+        
+        # Main data plots
+        plt.plot(compiled_binned_averages, compiled_binned_total_deviations, 
+                 label='Shot Noise + Read Noise')
+        # Assuming `lower_bound` is 6
+        plt.plot(compiled_binned_averages, 
+                 compiled_binned_shot_deviations, label='Shot Noise')
+        plt.errorbar(10**logged_averages, 10**logged_deviations, 
+                     yerr=10**(logged_deviations_error), color='gray', 
+                     label='Fitted Points')
+        
+        # Extra lines
+        plt.plot(10**full_range_averages, 
+                 10**(full_range_averages * 0.50 + parm1), 
+                 'k-', label='Forced 0.50 Gradient')
+        
+        # Reference lines and text annotations
+        plt.axhline(y=mean_rn_std_DN, color='k', linestyle='--')
+        plt.axvline(x=95000, color='k', linestyle='--')
+        plt.text(2, 1.5, f'Fitted k gain = {np.round(kgain,1)} e/DN', fontsize=15)
+        plt.text(2, 10, 
+                 f'Read noise (Gaussian fit) = {np.round(mean_rn_gauss_DN,1)} DN = {mean_rn_gauss_e} e', 
+                 fontsize=15)
+        plt.text(2, 8, f'Read noise (Std) = {np.round(mean_rn_std_DN,1)} DN = {np.round(mean_rn_std_e,1)} e', 
+                 fontsize=15)
+        
+        # Scaling and labeling
+        plt.xscale('log')
+        plt.yscale('log')
+        plt.xlabel('Signal (DN)')
+        plt.ylabel('Noise (DN)')
+        plt.title('PTC curve fit')
+        plt.grid(True)
+        
+        # Setting axis limits
+        plt.xlim([1, 20000])
+        plt.ylim([1, 1000])
+        
+        # Legend and aesthetics
+        plt.legend(loc='upper left')
+        
+        plt.gca().tick_params(axis='both', which='major', labelsize=18)
+        
+        # Adjust figure position
+        plt.gcf().set_dpi(100)
+        plt.gcf().set_size_inches(20, 10)
+        
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    # prepare PTC array
+    ptc_list = [compiled_binned_averages, 
+             compiled_binned_shot_deviations]
+    ptc = np.column_stack(ptc_list)
+
+    prhd, exthd = create_default_headers()
```

Comment:
You should grab a set of headers from the input data and add a history comment instead of filling the KGain object with mock headers.

Reply:
This was implemented by Juergen. I thought it had been reviewed. Sounds good.

Reply:
The input dataset has a set of frames, each has its own primary/extended header. The operations with the dataset are:

`dataset_nl.add_error_term(                dataset_nl.frames
dataset_nl.all_data                       dataset_nl.rescale_error(
dataset_nl.all_dq                         dataset_nl.save(
dataset_nl.all_err                        dataset_nl.split_dataset(
dataset_nl.copy(                          dataset_nl.update_after_processing_step(`

I do not see an option to collect all the information of the headers in a structured way. Is there one? That'd be great. If not, which object structure should the call to 
```
data.NonLinearityCalibration(nonlin_data,
        pri_hdr = prhd, ext_hdr = exthd, input_dataset=dataset_nl)
```
have when the dataset has a set of frames with different primary/extended headers? An example would help

Reply:
you could just take the headers from the first frame: 
`prhd = nonlin_data.frames[0].pri_hdr`
`exthd = nonlin_data.frames[0].ext_hdr`

Reply:
@maxwellmb If they are different for different frames as I wrote, how does this propagate each difference?

Reply:
Yeah, that's one of the downsides to this approach, but its better than having just dummy headers. Nobody should be using the GAIN keyword from a Non Linearity calibration anyway. Ultimately the files used will be kept track of (because we keep track of the input datasets), so if we ever need to trace it back its possible. The HISTORY keyword should be detailed enough to make it clear that the current fits file is a combination of several fits files.

Reply:
We could remove the useless headers too

Reply:
Done. Please check for any further suggestions

Reply:
Sure, deleting non-sense headers is definitely a nice-to-have to catch those subtle bugs (but naturally, low priority).